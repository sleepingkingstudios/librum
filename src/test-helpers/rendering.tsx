import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
