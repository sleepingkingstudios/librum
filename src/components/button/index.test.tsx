import * as React from 'react';

import '@testing-library/jest-dom';
import { faRadiation } from '@fortawesome/free-solid-svg-icons';
import userEvent from '@testing-library/user-event';

import { Button } from './index';
import { render } from '@test-helpers/rendering';

describe('<Button />', () => {
  it('should display the contents', () => {
    const { getByRole } = render(<Button>Greet</Button>);

    const button = getByRole('button');

    expect(button).toBeVisible();
    expect(button).toHaveTextContent('Greet');
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(<Button>Greet</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('with className: value', () => {
    const className = 'btn-custom';

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Button className={className}>Greet</Button>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with disabled: true', () => {
    it('should display the contents', () => {
      const { getByRole } = render(<Button disabled>Greet</Button>);

      const button = getByRole('button');

      expect(button).toBeVisible();
      expect(button).toHaveTextContent('Greet');
    });
  });

  describe('with htmlType: "button"', () => {
    const htmlType = 'button';

    it('should display the contents', () => {
      const { getByRole } = render(<Button htmlType={htmlType}>Greet</Button>);

      const button = getByRole('button');

      expect(button).toBeVisible();
      expect(button).toHaveTextContent('Greet');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button htmlType={htmlType}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with htmlType: "reset"', () => {
    const htmlType = 'reset';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button htmlType={htmlType}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with htmlType: "submit"', () => {
    const htmlType = 'reset';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button htmlType={htmlType}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with label: value', () => {
    const label = 'Greet';

    it('should display the contents', () => {
      const { getByRole } = render(<Button label={label} />);

      const button = getByRole('button');

      expect(button).toBeVisible();
      expect(button).toHaveTextContent('Greet');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button label={label} />);

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with icon: value', () => {
      const icon = faRadiation;

      it('should display the contents', () => {
        const { getByRole } = render(<Button icon={icon} label={label} />);

        const button = getByRole('button');

        expect(button).toBeVisible();
        expect(button).toHaveTextContent('Greet');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(<Button icon={icon} label={label} />);

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with onClick: value', () => {
    const onClick = jest.fn();

    beforeEach(() => { onClick.mockClear(); });

    it('should display the contents', () => {
      const { getByRole } = render(<Button onClick={onClick}>Greet</Button>);

      const button = getByRole('button');

      expect(button).toBeVisible();
      expect(button).toHaveTextContent('Greet');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button onClick={onClick}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });

    describe('when the button is clicked', () => {
      it('should call the function', async () => {
        const { getByRole } = render(<Button onClick={onClick}>Greet</Button>);

        const button = getByRole('button');

        await userEvent.click(button);

        expect(onClick).toHaveBeenCalled();
      });
    });

    describe('with disabled: true', () => {
      describe('when the button is clicked', () => {
        it('should not call the function', async () => {
          const { getByRole } = render(
            <Button disabled onClick={onClick}>Greet</Button>
          );

          const button = getByRole('button');

          await userEvent.click(button);

          expect(onClick).not.toHaveBeenCalled();
        });
      });
    });
  });

  describe('with outline: true', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Button outline>Greet</Button>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with size: lg', () => {
    const size = 'lg';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button size={size}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with size: sm', () => {
    const size = 'sm';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button size={size}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with type: "danger"', () => {
    const type = 'danger';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button type={type}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with outline: true', () => {
      it('should match the snapshot', () => {
        const { asFragment } = render(
          <Button outline type={type}>Greet</Button>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with type: "info"', () => {
    const type = 'info';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button type={type}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with outline: true', () => {
      it('should match the snapshot', () => {
        const { asFragment } = render(
          <Button outline type={type}>Greet</Button>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with type: "muted"', () => {
    const type = 'muted';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button type={type}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with outline: true', () => {
      it('should match the snapshot', () => {
        const { asFragment } = render(
          <Button outline type={type}>Greet</Button>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with type: "primary"', () => {
    const type = 'primary';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button type={type}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with outline: true', () => {
      it('should match the snapshot', () => {
        const { asFragment } = render(
          <Button outline type={type}>Greet</Button>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with type: "success"', () => {
    const type = 'success';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button type={type}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with outline: true', () => {
      it('should match the snapshot', () => {
        const { asFragment } = render(
          <Button outline type={type}>Greet</Button>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with type: "warning"', () => {
    const type = 'warning';

    it('should match the snapshot', () => {
      const { asFragment } = render(<Button type={type}>Greet</Button>);

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with outline: true', () => {
      it('should match the snapshot', () => {
        const { asFragment } = render(
          <Button outline type={type}>Greet</Button>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with url: value', () => {
    const url = 'path/to/resource';

    it('should display the contents', () => {
      const { getByRole } = render(
        <Button url={url}>Greet</Button>,
        { router: true },
      );

      const button = getByRole('link');

      expect(button).toBeVisible();
      expect(button).toHaveTextContent('Greet');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Button url={url}>Greet</Button>,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with disabled: true', () => {
      it('should display the contents', () => {
        const { getByRole } = render(
          <Button disabled url={url}>Greet</Button>,
          { router: true },
        );

        const button = getByRole('button');

        expect(button).toBeVisible();
        expect(button).toHaveTextContent('Greet');
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <Button disabled url={url}>Greet</Button>,
          { router: true },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });
});
