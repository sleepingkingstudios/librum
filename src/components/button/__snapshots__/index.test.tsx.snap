// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with className: value should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-custom"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with htmlType: "button" should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with htmlType: "reset" should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    type="reset"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with htmlType: "submit" should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    type="reset"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with label: value should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with label: value with icon: value should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-radiation pr-2"
      data-icon="radiation"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M215 186.7c-23.9 13.8-40 39.7-40 69.3L31 256c-17.7 0-32.2-14.4-30-31.9C9.7 154 46.8 92.7 100.3 52c14.1-10.7 33.8-5.3 42.7 10l72 124.7zM255 336c14.6 0 28.2-3.9 40-10.7l72 124.8c8.8 15.3 3.7 35.1-12.6 41.9c-30.6 12.9-64.2 20-99.4 20s-68.9-7.1-99.4-20c-16.3-6.9-21.4-26.6-12.6-41.9l72-124.8c11.8 6.8 25.4 10.7 40 10.7zm224-80l-144 0c0-29.6-16.1-55.5-40-69.3L367 62c8.8-15.3 28.6-20.7 42.7-10c53.6 40.7 90.6 102 99.4 172.1c2.2 17.5-12.4 31.9-30 31.9zM257 208a48 48 0 1 1 -2 96 48 48 0 1 1 2-96z"
        fill="currentColor"
      />
    </svg>
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with onClick: value should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with outline: true should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with size: lg should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-lg"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with size: sm should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-sm"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "danger" should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-danger"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "danger" with outline: true should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline btn-danger"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "info" should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-info"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "info" with outline: true should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline btn-info"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "muted" should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-muted"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "muted" with outline: true should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline btn-muted"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "primary" should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-primary"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "primary" with outline: true should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline btn-primary"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "success" should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-success"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "success" with outline: true should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline btn-success"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "warning" should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-warning"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with type: "warning" with outline: true should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline btn-warning"
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;

exports[`<Button /> with url: value should match the snapshot 1`] = `
<DocumentFragment>
  <a
    class="btn"
    href="/path/to/resource"
  >
    Greet
  </a>
</DocumentFragment>
`;

exports[`<Button /> with url: value with disabled: true should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    disabled=""
    type="button"
  >
    Greet
  </button>
</DocumentFragment>
`;
